sogs/model/message.py

```diff
@@ -0,0 +1,65 @@
+from .. import db
+from ..db import query
+
+from .user import User
+
+
+class Message:
+    """Class representing a DM between users
+
+    Properties:
+        sender: sender user of the message
+        recip: recipant user of the message
+        data: opaque message data
+        signature: signature of data
+    """
+
+    def __init__(self, row=None, *, sender=None, recip=None, data=None, signature=None):
+        """
+        Constructs a Message from a pre-retrieved row *or* sender recipiant data and signature.
+        """
+        if row is None:
+            if None in (sender, recip, data, signature):
+                raise ValueError(
+                    "Message() error: provided no row but did not provide all members for construction"
+                )
+            for arg in (sender, recip):
+                if not isinstance(arg, User):
+                    raise ValueError("Message() error: sender or recipient was not a User model")
+
+            with db.transaction():
+                row = query(
+                    "INSERT INTO inbox(sender, recipient, data, signature) VALUES(:sender, :recipient, :data, :signature) RETURNING *",
+                    sender=sender.id,
+                    recipiant=recip.id,
+                    data=data,
+                    signture=signature,
+                ).first()
+        # sanity check
+        assert row is not None
+        self._row = row
+
+    @property
+    def info(self):
+        return {
+            "data": self.data,
+            "signature": self.signature,
+            "sender": self.sender.session_id,
+            "recipient": self.recipient.session_id,
+        }
+
+    @property
+    def signature(self):
+        return self._row['signature']
+
+    @property
+    def data(self):
+        return self._row['data']
+
+    @property
+    def sender(self):
+        return User(id=self._row['sender'], autovivify=False)
+
+    @property
+    def recipient(self):
+        return User(id=self._row['recipient'], autovivify=False)
```

Comment:
Would be preferable to cache the `User` object here (in `self._sender` and `self._recipient` or something like that) so that if the caller writes code that does something like:
```Python
if msg.sender == whatever:
    print(msg.sender)
```
that you only fetch it from the database once.